plugin/manager.go

```diff
@@ -0,0 +1,185 @@
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"sync"
+	"time"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-multierror"
+	hplugin "github.com/hashicorp/go-plugin"
+)
+
+// PluginManager can start new plugins watch and kill all plugins.
+type PluginManager struct {
+	Interval time.Duration
+
+	sources     []withClient[Source]
+	destination []withClient[Destination]
+	m           sync.Mutex
+}
+
+// NewDestination returns a new Destination interface from a plugin path and configuration.
+func (pm *PluginManager) NewDestination(path string, config map[string]any) (Destination, error) {
+	pm.m.Lock()
+	defer pm.m.Unlock()
+
+	c := client(path)
+	cp, err := c.Client()
+	if err != nil {
+		c.Kill()
+		return nil, fmt.Errorf("failed to create rpc client interface: %w", err)
+	}
+	d, err := newDestination(cp)
+	if err != nil {
+		c.Kill()
+		return nil, err
+	}
+
+	if err := d.Configure(config); err != nil {
+		c.Kill()
+		return nil, fmt.Errorf("failed to configure destination: %w", err)
+	}
+
+	pm.destination = append(pm.destination, withClient[Destination]{t: d, c: c, path: path, config: config})
+
+	return d, nil
+}
+
+// NewSource returns a new Source interface from a plugin path and configuration.
+func (pm *PluginManager) NewSource(path string, config map[string]any) (Source, error) {
+	pm.m.Lock()
+	defer pm.m.Unlock()
+
+	c := client(path)
+	cp, err := c.Client()
```

Comment:
Right here we could cache the clients for sources (same for destinations) so that the watcher can use them